### PR TITLE
handling memcache failures

### PR DESCRIPTION
**Pull request by tzachyrm**
Some memcached backeds return `False` from the `add` function due to sporadic failures. In this case the following call to `incr` will raise a `ValueError`, and the call to `get` will return `None`.

### CODE REVIEW

File: ratelimit/utils.py [hunk 1]
```diff
@@ -153,9 +153,17 @@ def is_ratelimited(request, group=None, fn=None, key=None, rate=None,
         count = initial_value
     else:
         if increment:
-            count = cache.incr(cache_key)
+            try:
+                count = cache.incr(cache_key)
+            except ValueError: # add failed or value expired
```
Comment (willkg) [inline ratelimit/utils.py]: Nit: Comments require two space before them when they're inline like this.

File: ratelimit/utils.py [hunk 2]
```diff
@@ -153,9 +153,17 @@ def is_ratelimited(request, group=None, fn=None, key=None, rate=None,
...
+                cache.add(cache_key, initial_value)
+                count = initial_value
         else:
             count = cache.get(cache_key)
+            if count is None: # add failed or value expired
```
Comment (willkg) [inline ratelimit/utils.py]: Nit: Comments require two space before them when they're inline like this.

**Comment by willkg** [comment]
When you say "some memcached backends", which ones are you talking about specifically? Are there links to documentation about the issue?

File: ratelimit/utils.py (hunk 2 above)
Comment (jsocol) [inline ratelimit/utils.py]: You can run `./run.sh flake8` to run all the static analysis and style checks.

**Comment by adamchainz** [comment]
This affected us @YPlan too :smile: 

The cause of this is not 'sporadic', but a race condition with evictions; `add()` returns `False` since the key is there when it checks, but by the time `incr()` or `get()` runs, the key has been evicted due to other parallel traffic, or expired. This is only something you'd see with enough traffic.

The `ValueError`-on-failed-`incr()` behaviour is for all Django built-in memcached backends - the behaviour is normalized in `BaseMemcachedCache`:  https://github.com/django/django/blob/master/django/core/cache/backends/memcached.py#L105

We are currently using a similar patch like this one but with some tests :smile: I've submitted it as a second PR here: #81

**Comment by tzachyrm** [comment]
@willkg In my case it's the Google App Engine memcache. [The docs](https://cloud.google.com/appengine/docs/python/memcache/) state that:
> Memcache operations can fail for various reasons. Applications should be designed to catch failed operations without exposing these errors to end users. This applies especially to Set operations.

**Comment by jsocol** [comment]
Closing in favor of #81 (which itself may be fixed)

**Comment by jsocol** [comment]
@adamchainz et al, can you check to see if #88 fixed this for you?

**Comment by adamchainz** [comment]
Yes it did./

**Pull request closed without merging**